src/addrdb.cpp

```diff
@@ -114,16 +114,16 @@ bool DeserializeFileDB(const fs::path& path, Data& data, int version)
     FILE* file = fsbridge::fopen(path, "rb");
     CAutoFile filein(file, SER_DISK, version);
     if (filein.IsNull()) {
-        LogPrintf("Missing or invalid file %s\n", path.string());
+        LogPrintf("Missing or invalid file %s\n", fs::PathToString(path));
```

Comment:
Wasn't `path.string()` already explicit enough? Also, when should `u8string` be used instead? Finally, `strprintf("%s", path)` without any helpers still compiles? When should/shouldn't that be used?

Reply:
> Wasn't `path.string()` already explicit enough?

The problem with `path.string()` isn't the explicitness, it's the unsafe behavior with the new path class on windows, where it returns a string in an arbitrary encoding based on the local code page. In preparation for this, this PR is disabling the `path.string()` now so it isn't misused later, and adding `PathToString` as a replacement which preserves the behavior we need and is complementary to `PathFromStrong`.

> Also, when should `u8string` be used instead?

`u8string/u8path` functions should be used for converting strings/paths where the strings must be valid UTF-8 strings. `PathToString/PathFromString` functions should be used otherwise.

U8 functions are the right functions to call in JSON-RPC code and Qt string code because Qt strings and JSON requests/responses require valid UTF-8. But otherwise `PathToString/PathFromString` functions which pass along strings verbatim and avoid making encoding assumptions are more appropriate. `PathToString/PathFromString` functions are the right functions to call for parsing argv paths, parsing environment variable paths, parsing config paths, writing to log files, and calling POSIX APIs that expect and return native 8-bit path strings. In all these places UTF-8 encoding is typical and nice to have, but not something that should be enforced or required. If someone has their home directory set to a non-utf8 path, `PathToString/PathFromString` won't mangle the string or care about the encoding.

Everything above is actually documented in the followup commit 3e88c9f6b4e84ab5da78f4862266b5ff13bb8fc0 ([tag](https://github.com/ryanofsky/bitcoin/commits/review.20744.8-edit.1)) posted https://github.com/bitcoin/bitcoin/pull/20744#issuecomment-916627496, but I threw together the documentation really quickly while trying to converge on a solution, so it could use some cleanup and reshuffling, either in this PR or the std::filesystem PR later when these functions stop being identical stubs.

> Finally, `strprintf("%s", path)` without any helpers still compiles? When should/shouldn't that be used?

I can look into that because I'm not too familiar with tinyformat, but I would suspect that this always compiled and never worked at runtime maybe? I'm not sure how it could have worked unless tinyformat knew about boost types.

Reply:
I think tinyformat just "emulates" our `::ToString(path)`, which should also compile (didn't try)

Reply:
> I think tinyformat just "emulates" our `::ToString(path)`, which should also compile (didn't try)

Apparently tinyformat is using operator<<, which boost path and std path both provide. I found I could disable it with

```c++
template<class CharT, class Traits>
std::basic_ostream<CharT,Traits>& operator<<(std::basic_ostream<CharT,Traits>& os, const path& p)
{
    static_assert(sizeof(CharT) == -1, "Implicit path to string conversion is error-prone and not supported, please use explicit PathToString call");
    return os;
}
```

But I found that this also broke some boost tests, and also I'm a little reluctant to make this change anyway since formatting support seems convenient for debugging. I'm leaving this alone for now but happy to do whatever people prefer here.

Reply:
> PathToString/PathFromString functions should be used otherwise
> U8 functions are the right functions to call in JSON-RPC code and Qt string code ...


I am not sure how this can be achieved, without something like bilingual_str.

For example, `InitError`s are passed to Qt, as well as stderr, as well as the debug log. If Qt requires `u8string` and the others require `PathToString`, then with the current approach here this is impossible to achieve, no?

Reply:
Nothing we want to achieve should be impossible to achieve or require a different string type. If your home directory is `"/home/\xf0"` (invalid utf-8), and Qt needs to display an error message about peers.dat, it should just display it with a [replacement character](https://en.wikipedia.org/wiki/Specials_(Unicode_block)#Replacement_character) like "Error reading /home/�/.bitcoin/peers.dat". If we want to return the same file path as JSON-RPC, we might choose a different escaping mechanism like [python's lone surrogates](https://www.python.org/dev/peps/pep-0383/) `"/home/\udcf0/.bitcoin/peers.dat"`.

At the boundaries between our application and external programs and libraries like JSON-RPC and Qt, different ways of representing paths as strings will be necessary. That is why standard encoding functions like `u8path` and `u8string` are useful, and why we will choose different encoding functions in different places we are interacting with external code. But internally, any path can always be represented as a `path` object, and any path can always be represented as a string of bytes, as long as we are consistent about how we interpret those bytes. The `PathToString` and `PathFromString` function provide a consistent way to represent path objects as strings of bytes, now that we will no longer be able to rely on the `boost::filesystem::path` class's string methods. (The new `std::filesystem::path` class has similar methods, but unlike boost's methods their behavior cannot be controlled with an imbue function, so they are unsafe and expose complicated operating system specific behavior and will mangle strings, which is why this PR avoids and prevents calling them.)